s3/pkg/service/object.go

```diff
@@ -429,6 +397,30 @@ func (s *s3Service) CopyObject(ctx context.Context, in *pb.CopyObjectRequest, ou
 	srcObjectName := in.SrcObjectName
 	targetBucketName := in.TargetBucketName
 	targetObjectName := in.TargetObjectName
+	targetBucket, err := s.MetaStorage.GetBucket(ctx, targetBucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+
+	isAdmin, tenantId, _, err := util.GetCredentialFromCtx(ctx)
+	if err != nil {
+		log.Errorf("get credential faied, err:%v\n", err)
+		return err
+	}
+
+	if !isAdmin {
+		switch targetBucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if targetBucket.TenantId != tenantId {
+				err = ErrBucketAccessForbidden
+				return err
+			}
+		}
+	}
+
 	srcBucket, err := s.MetaStorage.GetBucket(ctx, srcBucketName, true)
```

Comment:
Is it necessary to check the access rights of source bucket and source object?

Reply:
@sfzeng  thanks, you are right, I will add the check in next commit